Playlist

LordAro opened the pull request:
Most definitely unfinished, nearly completely untested (hey, it compiles, :shipit:), but i might as well make a PRQ for review porpoises

<!---
@huboard:{"order":3.0,"milestone_order":3,"custom_state":""}
-->

LordAro committed:
Initial playlist work

LordAro committed:
Further playlist work, mostly dequeue

wlcx commented:
![image](https://cloud.githubusercontent.com/assets/3065381/6765374/6baf9898-cfd5-11e4-92a8-1c8a5135ab1c.png)
Review porpoise at the ready

LordAro committed:
Make hub less global

LordAro committed:
Add enqueue

LordAro committed:
Refactor to use AsSlice for arguments

LordAro committed:
Refactor request handler to use a function map

LordAro committed:
Add list request handler

LordAro committed:
Add select request handler

LordAro committed:
Stop 'bad' commands from getting through

LordAro committed:
Refactor some more, add some selection logic

LordAro committed:
Obiwan error

LordAro committed:
Readd list

LordAro committed:
Correctly fix the obiwan

wlcx committed:
Initialise playlist items to empty slice

wlcx committed:
Add initial tests

wlcx committed:
Fix test and add slightly more realistic data

LordAro committed:
Please god fix this obiwan error

wlcx committed:
Moar tests

wlcx committed:
Add testing for selection being adjusted

wlcx committed:
Rename functions for clarity

wlcx committed:
Use baps3.ServiceState for downstream service state

Update ServiceState on responses

wlcx committed:
Add dump request, dump complete state on client connect

wlcx committed:
Update to use baps3.State

LordAro committed:
Refactor to return slices of messages

wlcx committed:
Make all processReqX methods on hub

wlcx committed:
Move list and dump req handlers to own functions

This is because the list responses are used in dump, and the dump responses are sent when a client connects. It was inelegant to fake a request to generate this output.
Also, as a result dump now includes list

wlcx committed:
Remove unnecessary for loop

LordAro committed:
Merge pull request #7 from UniversityRadioYork/initial-dump

Add Dump

LordAro committed:
Change selection when adding/removing items

LordAro committed:
Send SELECT responses when the selection changes

LordAro committed:
Fix tests

wlcx committed:
Use Args() rather than AsSlice()

wlcx committed:
Move select load logic into processReqSelect, doesn't need special handling any more

wlcx committed:
Add autoadvance

Add advance method to playlist which sets selection to next available file
Handle RsEnd from downstream, conditionally advance on whether autoadvance enabled
Add processRqAutoadvance to allow toggling of autoadvance
Add playlist.Advance tests

wlcx committed:
s/Autoadvance/autoAdvance/, removal of multiple calls to Args(), tweaks

wlcx committed:
s/makeRsAutoadvance/makeRsAutoAdvance/ for consistency

wlcx committed:
Change playlist.Advance() to deselect when no next item

Slighltly neater implementation of Advance
Update tests

LordAro committed:
Merge pull request #8 from UniversityRadioYork/autoadvance

Add Autoadvance

LordAro committed:
Factor out bad command message
